power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -21,114 +25,242 @@ namespace detail = power_grid_model::optimizer::detail;
 
 struct TrafoGraphVertex {
-    bool is_source{}; // is_source = true if the vertex is a source
+    bool is_source{};
 };
 
 struct TrafoGraphEdge {
-    Idx2D pos{};
+    Idx2D regulated_idx{};
     EdgeWeight weight{};
+
+    bool operator==(const TrafoGraphEdge& other) const {
+        return regulated_idx == other.regulated_idx && weight == other.weight;
+    } // thanks boost
+
+    auto operator<=>(const TrafoGraphEdge& other) const {
+        if (auto cmp = weight <=> other.weight; cmp != 0) { // NOLINT(modernize-use-nullptr)
+            return cmp;
+        }
+        if (auto cmp = regulated_idx.group <=> other.regulated_idx.group; cmp != 0) { // NOLINT(modernize-use-nullptr)
+            return cmp;
+        }
+        return regulated_idx.pos <=> other.regulated_idx.pos;
+    }
+};
+
+using TrafoGraphEdges = std::vector<std::pair<TrafoGraphIdx, TrafoGraphIdx>>;
+using TrafoGraphEdgeProperties = std::vector<TrafoGraphEdge>;
+using RankedTransformerGroups = std::vector<Idx2D>;
+
+struct RegulatedObjects {
+    std::set<Idx> transformers{};
+    std::set<Idx> transformers3w{};
 };
 
-// TODO(mgovers): investigate whether this really is the correct graph structure
 using TransformerGraph = boost::compressed_sparse_row_graph<boost::directedS, TrafoGraphVertex, TrafoGraphEdge,
                                                             boost::no_property, TrafoGraphIdx, TrafoGraphIdx>;
 
+template <std::derived_from<ThreeWindingTransformer> Component, class ComponentContainer>
+    requires main_core::model_component_state_c<main_core::MainModelState, ComponentContainer, Component>
+constexpr void add_edge(main_core::MainModelState<ComponentContainer> const& state,
+                        RegulatedObjects const& regulated_objects, TrafoGraphEdges& edges,
+                        TrafoGraphEdgeProperties& edge_props) {
+    // Only add idx2d to one of the 2 directional edges of 3 winding transformer which ensures a single ranking
+    bool assigned_idx{};
+    using enum Branch3Side;
+    std::array<std::tuple<Branch3Side, Branch3Side>, 3> const branch3_combinations{
+        {{side_1, side_2}, {side_2, side_3}, {side_1, side_3}}};
+    for (auto const& transformer3w : state.components.template citer<ThreeWindingTransformer>()) {
+        for (auto const& [first_side, second_side] : branch3_combinations) {
+            if (!transformer3w.status(first_side) || !transformer3w.status(second_side)) {
+                continue;
+            }
+            auto const& from_node = transformer3w.node(first_side);
+            auto const& to_node = transformer3w.node(second_side);
+
+            auto const tap_at_first_side = transformer3w.tap_side() == first_side;
+            auto const single_direction_condition = regulated_objects.transformers3w.contains(transformer3w.id()) &&
+                                                    (tap_at_first_side || transformer3w.tap_side() == second_side);
+            if (single_direction_condition) {
```

Comment:
The suggestion makes it less descriptive.

Reply:
probably my confusion came from the fact that there's two logical steps in one statement

* it is regulated
* because of that, we model it as unidirectional

my guess is that one of my other comments on splitting up when the `contains` check happens should make this more clear